### PR TITLE
Netloc integration with mpich

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**
Features:
1. Integrate netloc topology query
2. network topology type classification
3. comm_split_type with hints: switch levels in tree topology

### CODE REVIEW

File: src/mpid/ch3/channels/nemesis/src/ch3_init.c [hunk 1]
```diff
@@ -44,6 +44,9 @@ static int split_type(MPIR_Comm * user_comm_ptr, int stype, int key,
          * to the upper layer */
         mpi_errno = MPIR_Comm_split_type(comm_ptr, stype, key, info_ptr, newcomm_ptr);
         goto fn_exit;
+    } else if (stype == MPIX_COMM_TYPE_NEIGHBORHOOD) {
+        mpi_errno = MPIR_Comm_split_type_neighborhood(comm_ptr, stype, key, info_ptr, newcomm_ptr);
+        goto fn_exit;
```
Comment (pavanbalaji) [inline src/mpid/ch3/channels/nemesis/src/ch3_init.c]: You'll never hit this `else if` branch.

File: src/include/netloc_util.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#ifndef NETLOC_UTIL_H_INCLUDED
+#define NETLOC_UTIL_H_INCLUDED
+
+#include "netloc.h"
+
+typedef enum {
+    MPIR_NETLOC_NETWORK_TYPE__FAT_TREE,
+    MPIR_NETLOC_NETWORK_TYPE__CLOS_NETWORK,
+    MPIR_NETLOC_NETWORK_TYPE__INVALID,
+} MPIR_Netloc_network_type;
```
Comment (pavanbalaji) [inline src/include/netloc_util.h]: This should be renamed to `MPIR_Netloc_network_topo_type`.

File: src/include/netloc_util.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+             * the network */
+            int *node_levels;
+        } tree;
+    } u;
+} MPIR_Netloc_network_attributes;
+
+int MPIR_Netloc_parse_topology(netloc_topology_t topology,
+                               MPIR_Netloc_network_attributes * network_attr);
+
+int MPIR_get_network_end_point(MPIR_Netloc_network_attributes, netloc_topology_t netloc_topology,
```
Comment (pavanbalaji) [inline src/include/netloc_util.h]: First character after `MPIR_` should be capitalized.

**Comment by KavithaTipturMadhu** [comment]
@pavanbalaji I have updated the patches as per review comments.

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
@raffenet @pavanbalaji Barring one freebsd test failing because of slurm, the tests are green and I have addressed all the comments. Could you guys give it a look?

**Comment by pavanbalaji** [comment]
I've signed off on the first commit.  For the second commit, I made some cleanup changes and signed off on it.  Please take a look.

I'm not particularly happy with the third commit.  What's the point of the `node_index` structure?  It doesn't seem to be used anywhere.  Also, this patch is mixing the network type (e.g., InfiniBand) with the network topology (e.g., fat tree).  One can create a fat tree topology with other network types too.  I understand that you need some IB-specific code to figure out the edge weights (e.g., SDR, DDR, QDR), but I think this needs to be hidden inside netloc.  For example, can you have netloc add the "edge weight" for each network edge with the link bandwidth?  Then MPICH will not need any IB-specific code at the MPI layer and can simply rely on the edge weights to build the topology.

**Comment by KavithaTipturMadhu** [comment]
Replying to your previous comment @pavanbalaji, the node_index struct was for torus network, I forgot to remove it. I'll remove that and double check for stray code. I have addressed the other comment to add bandwidth in netloc and used it. I have committed my netloc changes [here](https://github.com/pmodels/netloc/commit/47c6847c17a3fab86c7305b82309e3d69098c3e2)

**Comment by raffenet** [comment]
@KavithaTipturMadhu should I review this or are the other comments still being addressed?

**Comment by KavithaTipturMadhu** [comment]
@raffenet I have addressed the comments.

File: src/include/netloc_util.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+typedef enum {
+    MPIR_NETLOC_NETWORK_TYPE__FAT_TREE,
+    MPIR_NETLOC_NETWORK_TYPE__CLOS_NETWORK,
+    MPIR_NETLOC_NETWORK_TYPE__INVALID,
+} MPIR_Netloc_network_topo_type;
+
+typedef struct {
+    netloc_node_t *node;
+    long index;
+} node_index;
```
Comment (pavanbalaji) [inline src/include/netloc_util.h]: I still see `node_index`.  Please really fix my comments.

Comment (KavithaTipturMadhu) [reply src/include/netloc_util.h]: I am so sorry about that, I have fixed this.

**Comment by KavithaTipturMadhu** [comment]
@raffenet @pavanbalaji I had missed one of the changes, I have pushed it now, sorry about that.

File: src/include/netloc_util.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        } tree;
+    } u;
+} MPIR_Netloc_network_attributes;
+
+int MPIR_Netloc_parse_topology(netloc_topology_t topology,
+                               MPIR_Netloc_network_attributes * network_attr);
+
+int MPIR_Get_network_end_point(MPIR_Netloc_network_attributes, netloc_topology_t netloc_topology,
+                               hwloc_topology_t hwloc_topology, hwloc_cpuset_t bindset,
+                               netloc_node_t ** end_point);
```
Comment (pavanbalaji) [inline src/include/netloc_util.h]: The function names are not consistent.  If it's related to netloc, can you use the `MPIR_Netloc_` prefix?

File: src/mpi/comm/comm_split_type.c [hunk 6]
```diff
@@ -497,6 +619,65 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
...
+        goto use_node_comm;
+
+    hintval_global = (char *) MPL_malloc(sizeof(hintval), MPL_MEM_OTHER);
+    /* Compare the hintval strings */
+    mpi_errno = compare_info_args(hintval, hintval_global,
+                                  (strlen(hintval) + 1) * sizeof(char), comm_ptr,
+                                  &info_args_are_equal);
+
+    if (hintval_global != NULL)
+        MPL_free(hintval_global);
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: The checks to make sure the info hints are the same for all processes are being duplicated.  They need to be consolidated into one place.

**Comment by pavanbalaji** [comment]
@raffenet I've not reviewed the algorithms in the above commits.  Can you do that?

File: src/include/netloc_util.h (hunk 5 above)
Comment (KavithaTipturMadhu) [reply src/include/netloc_util.h]: This function is a utility I wrote to get netloc node for the current process. Does this name sound right? MPIR_Netloc_get_network_end_point

**Comment by raffenet** [comment]
Sure, I can review the algorithms.

File: src/mpi/init/netloc_util.c [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
+        while (!netloc_lookup_table_iterator_at_end(hti)) {
+            netloc_node_t *node = (netloc_node_t *) netloc_lookup_table_iterator_next_entry(hti);
+            int num_edges;
+            netloc_edge_t **edges;
+
+            if (node == NULL) {
+                break;
+            }
+            netloc_get_all_edges(topology, node, &num_edges, &edges);
+            if (num_edges > 0) {
```
[the remark this answers is not in the document]
Comment (KavithaTipturMadhu) [reply src/mpi/init/netloc_util.c]: Removed the check

File: src/mpi/init/netloc_util.c [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+                if (num_parents == num_parents_covered) {
+                    int added_previously = 0;
+                    for (k = 0; k < end_index; k++) {
+                        if (traversal_order[k] == neighbor) {
+                            added_previously = 1;
+                            break;
+                        }
+                    }
+                    if (!added_previously) {
+                        traversal_order[end_index++] = neighbor;
```
[the remark this answers is not in the document]
Comment (KavithaTipturMadhu) [reply src/mpi/init/netloc_util.c]: Once add nodes are added, there is no more insertion. I have added assert to ensure end index is less than the number of nodes. Do you think that would suffice?

File: src/mpi/init/netloc_util.c [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+
+                /* Mark the host node as visited */
+                network_attr->u.tree.node_levels[topology->nodes[i]->__uid__] = 0;
+                visited_count++;
+
+                /* Copy all parents without duplicates to the traversal list */
+                netloc_get_all_edges(topology, topology->nodes[i], &num_edges, &edges);
+                for (j = 0; j < num_edges; j++) {
+                    if (network_attr->u.tree.node_levels[edges[j]->dest_node->__uid__] < 0) {
+                        traversal_order[traversal_end++] = edges[j]->dest_node;
```
[the remark this answers is not in the document]
Comment (KavithaTipturMadhu) [reply src/mpi/init/netloc_util.c]: I have moved the initialization statements to before the loop begins.

**Comment by KavithaTipturMadhu** [comment]
I can attach a couple of raw netloc json files here to test manually but I don't have any automated positive tests for this. I have addressed other refactoring comments from Pavan.

File: src/mpi/init/netloc_util.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpi/init/netloc_util.c]: Yes, assert is a good idea. Thanks.

**Comment by raffenet** [comment]
OK, cool. I think I have a decent understanding of how the detection algorithms are supposed to work. I'm OK with moving forward with this.

**Comment by KavithaTipturMadhu** [comment]
@raffenet I have attached here a couple of test json files here, mostly boundary conditions and some positive tests too. I made a couple of minor changes in code and would restart the tests. Could you take a look at the changes for me? Thanks.
[graph_examples.zip](https://github.com/pmodels/mpich/files/2107597/graph_examples.zip)

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/comm/comm_split_type.c [hunk 10]
```diff
@@ -393,6 +393,38 @@ static int compare_info_args(const void *sendbuf, void *recvbuf, int count, MPIR
...
+static int compare_info_hint(const char *hintval, MPIR_Comm * comm_ptr, int *info_args_are_equal)
+{
+    int hintval_size = strlen(hintval);
+    int hintval_size_global;
+    char *hintval_global = NULL;
+    int mpi_errno = MPI_SUCCESS;
+
+    /* Compare hintval string length */
+    mpi_errno =
+        reduce_info_hint(&hintval_size, &hintval_size_global, sizeof(int) / sizeof(char), comm_ptr,
```
Comment (raffenet) [inline src/mpi/comm/comm_split_type.c]: The implementation of `reduce_info_hint` still seems a bit overly complex, but I think we can solve that later.

File: src/mpi/comm/comm_split_type.c [hunk 11]
```diff
@@ -580,14 +759,8 @@ int MPIR_Comm_split_type_neighborhood(MPIR_Comm * comm_ptr, int split_type, int
...
 #endif
-    } else {
-        /* In the mean time, the user passed in
-         * COMM_TYPE_NEIGHBORHOOD but did not give us an info we
-         * know how to work with.  Throw up our hands and treat it
-         * like UNDEFINED.  This will result in MPI_COMM_NULL
-         * being returned to the user. */
-        *newcomm_ptr = NULL;
-    }
+    } else
```
Comment (raffenet) [inline src/mpi/comm/comm_split_type.c]: A style nitpick - I prefer to leave the braces here. If we use braces for one part of a conditional, we should use them for all parts.

File: src/mpi/comm/comm_split_type.c [hunk 12]
```diff
@@ -507,6 +629,63 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
...
+        mpi_errno = network_split_switch_level(comm_ptr, key, switch_level, newcomm_ptr);
+        goto fn_exit;
+    } else if (!strncmp(hintval, "subcomm_min_size:", strlen("subcomm_min_size:"))
+               && *(hintval + strlen("subcomm_min_size:")) != '\0') {
+        int size = atoi(hintval + strlen("subcomm_min_size:"));
+        mpi_errno = network_split_by_size(comm_ptr, key, size, newcomm_ptr);
+        goto fn_exit;
+    }
+#endif
+  use_node_comm:
```
Comment (raffenet) [inline src/mpi/comm/comm_split_type.c]: Is this label accurate? It seems we are just returning a `NULL` comm and not doing anything with the node comm.

Comment (KavithaTipturMadhu) [reply src/mpi/comm/comm_split_type.c]: Changed the name of the label

File: src/mpi/comm/comm_split_type.c (hunk 11 above)
Comment (KavithaTipturMadhu) [reply src/mpi/comm/comm_split_type.c]: Fixed this

File: src/mpi/comm/comm_split_type.c [hunk 13]
```diff
@@ -12,6 +12,11 @@
 #include "hwloc.h"
 #endif
 
+#ifdef HAVE_NETLOC
+#include "netloc_util.h"
+#include <math.h>
```
Comment (raffenet) [inline src/mpi/comm/comm_split_type.c]: What's `math.h` for in this case?

File: src/mpi/comm/comm_split_type.c [hunk 14]
```diff
@@ -493,6 +627,63 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
...
+        mpi_errno = network_split_switch_level(comm_ptr, key, switch_level, newcomm_ptr);
+        goto fn_exit;
+    } else if (!strncmp(hintval, "subcomm_min_size:", strlen("subcomm_min_size:"))
+               && *(hintval + strlen("subcomm_min_size:")) != '\0') {
+        int size = atoi(hintval + strlen("subcomm_min_size:"));
+        mpi_errno = network_split_by_size(comm_ptr, key, size, newcomm_ptr);
+        goto fn_exit;
+    }
+#endif
+  use_null_comm:
```
Comment (raffenet) [inline src/mpi/comm/comm_split_type.c]: It might look better to set to `NULL` at the beginning of the function and then overwrite it later when conditions are met. Personally, all these `goto`s make me a little uncomfortable.

File: src/mpi/comm/comm_split_type.c (hunk 13 above)
Comment (KavithaTipturMadhu) [reply src/mpi/comm/comm_split_type.c]: I had added it for torus split, removed it for this patch.

**Comment by raffenet** [comment]
OK, I think my comments have been addressed. I have a couple minor fixups stashed locally (mostly to commit messages). @pavanbalaji should I signoff and merge?

**Comment by pavanbalaji** [comment]
I'd recommend removing `min_subcomm_size` for now, till that's ready.  The current implementation for that is incorrect.  That can go into the next PR.

**Comment by KavithaTipturMadhu** [comment]
@raffenet @pavanbalaji  Removed min_subcomm_size hint support, creating a new PR to support this.

File: src/mpi/comm/comm_split_type.c (hunk 10 above)
Comment (pavanbalaji) [reply src/mpi/comm/comm_split_type.c]: Do we need the `reduce_info_hint` function?  Can't that just be embedded into this function since this is the only place that it's used?

File: src/mpi/comm/comm_split_type.c [hunk 15]
```diff
@@ -505,6 +589,57 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
...
+        MPIR_ERR_POP(mpi_errno);
+
+    /* if all processes have the same hintval, perform
+     * topology-aware comm split */
+    if (info_args_are_equal) {
+#ifdef HAVE_HWLOC
+        /* if our bindset is not valid, skip topology-aware comm split */
+        if (!MPIR_Process.bindset_is_valid)
+            goto fn_exit;
+#endif
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: Is this needed for splitting by network topology?

File: src/mpi/comm/comm_split_type.c [hunk 16]
```diff
@@ -579,14 +714,8 @@ int MPIR_Comm_split_type_neighborhood(MPIR_Comm * comm_ptr, int split_type, int
         *newcomm_ptr = dummycomm_ptr;
 #endif
     } else {
-        /* In the mean time, the user passed in
-         * COMM_TYPE_NEIGHBORHOOD but did not give us an info we
-         * know how to work with.  Throw up our hands and treat it
-         * like UNDEFINED.  This will result in MPI_COMM_NULL
-         * being returned to the user. */
-        *newcomm_ptr = NULL;
+        MPIR_Comm_split_type_network_topo(comm_ptr, key, info_ptr, newcomm_ptr);
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: This doesn't sound right.  If the user didn't give a valid info key, we call `MPIR_Comm_split_type_network_topo`?

File: src/mpi/comm/comm_split_type.c [hunk 17]
```diff
@@ -579,14 +714,8 @@ int MPIR_Comm_split_type_neighborhood(MPIR_Comm * comm_ptr, int split_type, int
...
     }
-
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: White space change.

**Comment by pavanbalaji** [comment]
Some of the commit messages need changes based on the removal of `subcomm_min_size`.

**Comment by KavithaTipturMadhu** [comment]
@raffenet @pavanbalaji I have addressed all the comments, starting tests again.

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
You ignored this comment: `Some of the commit messages need changes based on the removal of subcomm_min_size.`

**Comment by KavithaTipturMadhu** [comment]
@pavanbalaji sorry I missed that change, I have updated it. I also removed your sign-off as suggested.

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/comm/comm_split_type.c [hunk 18]
```diff
@@ -542,42 +542,101 @@ int MPIR_Comm_split_type(MPIR_Comm * user_comm_ptr, int split_type, int key,
...
+    hintval_size = strlen(hintval);
+    mpi_errno =
+        MPIR_Allreduce(&hintval_size, &hintval_size_max, 1, MPI_INT, MPI_MAX, comm_ptr, &errflag);
+
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    /*Atleast one of the processes has "nbhd_common_dirname" hint */
+    if (hintval_size_max) {
+        MPIR_Comm_split_type_nbhd_common_dir(comm_ptr, key, hintval, newcomm_ptr);
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: I'm not sure I understand this.  Why is this being called when `hintval_size_max != 0`?

**Comment by KavithaTipturMadhu** [comment]
@pavanbalaji I have updated the pr as per your comments.

**Comment by raffenet** [comment]
is this ready to merge?

**Comment by KavithaTipturMadhu** [comment]
@raffenet i updated the patches with Pavan's comments. @pavanbalaji  do you want to take a look?

File: src/mpi/comm/comm_split_type.c [hunk 19]
```diff
@@ -366,27 +366,68 @@ static int node_split_gpu_device(MPIR_Comm * comm_ptr, int key,
 
 static const char *SHMEM_INFO_KEY = "shmem_topo";
 
-static int compare_info_args(const void *sendbuf, void *recvbuf, int count, MPIR_Comm * comm_ptr,
-                             int *info_args_global)
+static int compare_info_hint(const char *hintval, MPIR_Comm * comm_ptr,
+                             int *arg_info_args_are_equal)
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: What's with the variable name?

File: src/mpi/comm/comm_split_type.c [hunk 20]
```diff
@@ -542,42 +542,97 @@ int MPIR_Comm_split_type(MPIR_Comm * user_comm_ptr, int split_type, int key,
...
+    int info_args_are_equal;
+    int hintval_size, hintval_size_max;
+    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+
+    *newcomm_ptr = NULL;
+
+    /* We plan on dispatching different NEIGHBORHOOD support to
+     * different parts of MPICH, based on the key provided in the
+     * info object.  Right now, the one NEIGHBORHOOD we support is
+     * "nbhd_common_dirname", implementation of which lives in ROMIO */
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: This comment is no longer relevant.

File: configure.ac [hunk 21]
```diff
@@ -1514,6 +1514,43 @@ HWLOC_DO_AM_CONDITIONALS
...
+            [AS_HELP_STRING([[--with-netloc-prefix[=DIR]]],
+                            [use the NETLOC library installed in DIR]) or sys to use the system library], [],
+                            [with_netloc_prefix=no])
+
+netloclibdir=""
+AC_SUBST([netloclibdir])
+
+if test "$have_hwloc" = "yes" ; then
+    if test "${with_netloc_prefix}" != "no" ; then
+        if test "${with_netloc_prefix}" != "sys"; then
```
Comment (pavanbalaji) [inline configure.ac]: Are we using `sys` for system-installed netloc instead of `system`?  Is that consistent with other uses?

File: configure.ac [hunk 22]
```diff
@@ -1514,6 +1514,43 @@ HWLOC_DO_AM_CONDITIONALS
...
+                PAC_APPEND_FLAG([-L${with_netloc_prefix}/lib],[WRAPPER_LDFLAGS])
+                netloclibdir="-L${with_netloc_prefix}/lib"
+            fi
+       else
+           AC_LINK_IFELSE([AC_LANG_PROGRAM([#include "netloc.h"
+                                   ],
+                                   [])],
+                          [AC_MSG_ERROR([the Netloc installation seems to be added from system path])], [])
+        fi
+        AC_DEFINE(HAVE_NETLOC,1,[Define if netloc is available])
```
Comment (pavanbalaji) [inline configure.ac]: `HAVE_NETLOC` is not defined for system-installed netloc?

**Comment by pavanbalaji** [comment]
Please have a prefix for each commit.

**Comment by pavanbalaji** [comment]
You can also add my signoffs on the last two commits.

File: configure.ac (hunk 22 above)
Comment (KavithaTipturMadhu) [reply configure.ac]: AC_DEFINE is after the if else block that checks for netloc, but within the `-with-netloc-prefix!=no` check. Hence, it will be defined for either system defined netloc or user specified path.

File: configure.ac (hunk 21 above)
Comment (KavithaTipturMadhu) [reply configure.ac]: Sorry I missed this, updated to `system`.

File: src/mpi/comm/comm_split_type.c (hunk 20 above)
Comment (KavithaTipturMadhu) [reply src/mpi/comm/comm_split_type.c]: Removed.

**Comment by pavanbalaji** [comment]
The last commit should have "test" as the prefix

**Comment by KavithaTipturMadhu** [comment]
@pavanbalaji I have addressed your comments.

File: src/mpi/comm/comm_split_type.c [hunk 23]
```diff
@@ -366,27 +366,67 @@ static int node_split_gpu_device(MPIR_Comm * comm_ptr, int key,
...
+static int compare_info_hint(const char *hintval, MPIR_Comm * comm_ptr, int *info_args_are_equal)
 {
+    int hintval_size = strlen(hintval);
+    int hintval_size_global;
+    char *hintval_global = NULL;
     int mpi_errno = MPI_SUCCESS;
     MPIR_Errflag_t errflag = MPIR_ERR_NONE;
-    int info_args_are_equal = 0;
+    int info_args_global = 0;
+    int info_args_equal = 0;
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: Please use more descriptive variable names.

**Comment by pavanbalaji** [comment]
You can add my signoff on `comm_split_type: refactoring comm_split_type_neighborhood`

**Comment by pavanbalaji** [comment]
This commit needs a more detailed description: `test: update for nhbd matching info hint`

**Comment by pavanbalaji** [comment]
This commit needs a more detailed description: `netloc: topology discovery and attributes`

**Comment by pavanbalaji** [comment]
This commit has a totally random description.  Please be more careful when writing commit messages:

```
commit d70b41174d76bab0e51110871a37246b0720b959
Author: Kavitha Tiptur Madhu <kmadhu@anl.gov>
Date:   Tue May 22 22:24:34 2018 -0500

    comm_split_type: split by switch level for trees
    
    Signed-off-by: Pavan Balaji <balaji@anl.gov>
    
    comm split fixup
```

File: src/mpi/comm/comm_split_type.c (hunk 23 above)
Comment (KavithaTipturMadhu) [reply src/mpi/comm/comm_split_type.c]: I have renamed the variables and added a comment describing them

**Comment by KavithaTipturMadhu** [comment]
Verified correctness of changes pushed by @pavanbalaji and merged into `comm_split_type: refactoring info argument comparison` patch.

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by KavithaTipturMadhu** [comment]
Two tests are failing because of slurm issues in freebsd. I checked the failing ucx test manually, seems to be working. Restarting ucx tests.

**Comment by KavithaTipturMadhu** [comment]
test:jenkins/ch4/ucx

**Comment by KavithaTipturMadhu** [comment]
Two tests in ucx were failing because of PMIx patches, informed @raffenet.
